File: app/src/main/java/uk/co/thomasrynne/swisher/core/MediaHandler.scala
package uk.co.thomasrynne.swisher.model

import android.net.Uri
import org.json.JSONObject

/**
 */
trait MediaHandler {
  def handle(json: JSONObject):HandleResult
}

trait HandleResult
case object NotApplicableHandleResult extends HandleResult
case object NotAvailableHandleResult extends HandleResult
case class SuccessHandleResult(tracks:Tracks) extends HandleResult

trait PlayerProgress
case object UnknownPlayerProgress extends PlayerProgress
case class KnownPlayerProgress(total:Int, progress:Int) extends PlayerProgress

trait PlayerListener {
  def finished()
  def playing()
  def paused()
  def stopped()
  def onTrack(position:Int)
  def updateProgress(progress:PlayerProgress)
}
trait PlayerInit
case object CuePlayerInit extends PlayerInit
case object PlayPlayerInit extends PlayerInit
case class PlayTrackInit(track:Int) extends PlayerInit

trait TrackDescription {
  def name: String
  def image: Option[Uri]
}
trait Tracks {
  def name: String
  def thumbnail: Option[Uri]
  def tracks: List[TrackDescription]
  def player(init:PlayerInit, listener:PlayerListener):TracksPlayer
}

File: app/src/main/java/thomas/swisher/MediaHandler.java
package thomas.swisher;

import android.net.Uri;

import com.google.common.base.Optional;

import java.util.List;

import lombok.Value;
import thomas.swisher.shared.Core;
import thomas.swisher.utils.Utils;
import thomas.swisher.service.player.TracksPlayer;

/**
 */
public interface MediaHandler {

    public Optional<PlaylistEntry> handle(Utils.FlatJson json);

    public enum PlayerNotification { Paused, Playing, Finished }
    public interface PlayerListener {
        //Informs the main player that this player has paused/playing/finished all tracks
        public void notify(PlayerNotification notification);
        //Informs the main player that this player has moved onto the next track in its list
        public void onTrack(boolean autoPlayedThisTrack, int position);
        //Updates the main player with the current progress of the track
        //This should be called each time a new track starts and on pause/play events
        //The ui will assume that when a track is playing progress proceeds at 1ms per ms ;)
        public void currentProgress(Core.PlayerProgress progress);
    }

    public interface ThePlayer {
        public TracksPlayer create(boolean playNow, int currentTrack, boolean playNext, PlayerListener listener);
    }

    @Value
    public static class PlaylistEntry {
        public String name;
        public Optional<Uri> thumbnail;
        public List<? extends TrackDescription> tracks;
        public ThePlayer player;
    }

    public interface TrackDescription {
        public String name();
        public Optional<Uri> image();
    }

}
